spring worker version

OlexTul opened the pull request:
Pull-request contains spring **worker** version.

IfLunatic committed:
refactored pom.xml, dropped .properties, ScenarioSourceListenerImpl.java, ProxySourceClientImpl.java, JSONFileProxyProvider.java, JSONFileScenarioProvider.java, added V1__init.sql

IfLunatic committed:
refactored Scenario.java, SourceListenerData.java, ThreadPoolExecutorConfig.java, WebDriverConfig.java

git-alexbondarenko committed:
- add DTO to exchange data with client;
- drop PropertiesConstants.java

OlexTul committed:
- added ScenarioSourceController.java, Routes.java, SourceHandlerImpl.java;
- refactored ScenarioSourceQueueHandler, ProxySourceQueueHandler, ScenarioSourceQueue, ProxySourceQueue, pom.xml

git-alexbondarenko committed:
- refactor WebDriverInitializerImpl;
- drop WebDriverConfig;
- add StepResult DTO;

OlexTul committed:
- refactored ScenarioSourceController.java (ScenarioResultResponse); ScenarioResultResponse.java;
- dropped SourceListenerData.java, SourceListenerDataTest.java

git-alexbondarenko committed:
minor changes in StepResult DTO

Mercurialf committed:
Added ScenarioResultService, StepExecutionLoggingService, ScenarioResultServiceImpl, StepExecutionLoggingServiceImpl, ScenarioResultRepository, StepExecutionLoggingRepository, refactored V1_init.sql.

Mercurialf committed:
Added new Repository for StepResult, ScenarioResult

Mercurialf committed:
Refactored StepResult

Mercurialf committed:
Added ScenarioResultLoggingService, ScenarioResultLoggingServiceImpl. Update ScenarioExecutorImpl.

OlexTul committed:
- added create, get, delete options to ScenarioSourceController.java, ScenarioResultExceptions.java;
- refactored SourceHandlerImpl.java, ProxySourceQueueHandler.java, ScenarioResult.java, ScenarioResultResponse.java, V1__init.sql

OlexTul committed:
- refactored ParallelFlowExecutorServiceTest.java

OlexTul committed:
- added ParallelFlowExecutorServiceController.java;
- refactored JavaDoc in ScenarioSourceController.java

OlexTul committed:
- refactored ScenarioResultResponse.java (added StepResultResponse.java)

git-alexbondarenko committed:
add maven assembly plugin

OlexTul committed:
- added logging to the console in logback.xml

jurok3x commented on src/main/java/executor/service/controller/ScenarioSourceController.java:
```diff
@@ -0,0 +1,80 @@
+@RequestMapping(value = SCENARIOS)
+public class ScenarioSourceController {
+
+    private final SourceHandler handler;
+    private final ScenarioResultLoggingService scenarioResultLoggingService;
```
2 different services per 1 controller is not good

OlexTul replied:
It is good idea. I did it. Commit 1cc61c2a.

jurok3x commented on src/main/java/executor/service/model/request/ScenarioRequest.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public record ScenarioRequest(
+
+        @Valid
```
@NotEmpty is better

OlexTul replied:
1. Using @Valid:
@Valid specifies that the scenarios field must pass the validation defined in the Scenario class (and its sub-objects, if any).
This option is useful if you need to ensure that each element in the list of scenarios is also validated according to the rules defined in the Scenario class.
This is suitable if you want to ensure that all scenarios in the list comply with your business rules.

2. Using @NotEmpty:
@NotEmpty ensures that the scenarios field is not empty (that is, the list is not null or empty).
This option is useful if you simply care that the list of scripts is provided and not empty.
However, it does not validate the contents of the list, so the internal details of each script may not meet your requirements.

Conclusion: for our tasks it is necessary to use @Valid.

jurok3x commented on src/main/java/executor/service/model/Step.java:
```diff
@@ -1,9 +1,24 @@
+
 import java.util.Objects;
 
+@Entity
+@Table(name = "steps")
 public class Step {
```
Persistence object and DTO are 2 different things.

OlexTul replied:
I agree with you and fixed it. Commit f05e7d0a.

jurok3x commented on src/main/java/executor/service/service/impl/ScenarioResultLoggingServiceImpl.java:
```diff
@@ -0,0 +1,129 @@
+@Service
+public class ScenarioResultLoggingServiceImpl implements ScenarioResultLoggingService {
+
+    private final ScenarioResultRepository scenarioResultRepository;
+    private final StepResultRepository stepResultRepository;
+    private final StepRepository stepRepository;
```
Here is definitely something wrong with dependencies. 3 dependencies per service.

OlexTul replied:
The main idea was to work with ScenarioResult, namely to write into the database, the entities: StepResult, Step - necessary for normalizing the database, i.e. according to application requirements - there is no need to separately code API for StepResult and Step. If you have specific recommendations, please write.
Additionally, please indicate, if necessary, an API for StepResult and Step - a contract between the Client and the Worker, where this circumstance will be indicated.

Sokol3783 commented on src/main/java/executor/service/config/bean/BeanConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+ * @version 01
+ */
+@Configuration
+public class BeanConfiguration {
```
it's useless name witch describe anything or too common things, you will collect all bean configurations in this class? You will mix all layer's configurations  in one class? Or you wouldn't divide configurations on task area?
You decided don't use another cases of decomposition's strategies?

OlexTul replied:
Separating into separate classes is a good idea. I did it. Commit f05e7d0a.

Sokol3783 commented on src/main/resources/db/migration/V1__init.sql:
```diff
@@ -0,0 +1,33 @@
+    scenario_result_id bigint not null,
+    step_result_id     bigint not null,
+    primary key (scenario_result_id, step_result_id),
+    constraint scenario_step_scenario_fk foreign key (scenario_result_id)
+        references scenario_results (id) on delete cascade,
+    constraint scenario_step_step_fk foreign key (step_result_id)
+        references step_results (id) on delete cascade
```
I think it better setting foreign keys after initialization in all tables.

OlexTul replied:
Setting foreign keys after tables are initialized is a valid approach that can have certain advantages, especially when creating a database schema. It allows more flexible control over the order in which tables are initialized and can make data imports or migrations easier.

However, there are a few important points to consider:
1. Approach to migrations: If you plan to use database migrations such as Flyway or Liquibase, setting foreign keys should be part of the migration scripts. This ensures that the database schema is created and modified consistently according to your specification.
2. Data Security: Setting foreign keys after tables have been initialized can increase the risk of dirty data (data that does not comply with foreign key constraints) if data was inserted before the constraints were set. This can lead to undesirable consequences and difficulties in the future.
3. Data Dependencies: If your database design involves circular dependencies between tables, this may require more careful management of the order in which foreign keys are installed. In some cases, it may be necessary to temporarily disable or break foreign keys to insert initial data.

In general, when to install foreign keys depends on the specific requirements of your application and database methodology. Unless you have a specific reason for setting foreign keys after tables are initialized, it is generally recommended to include them in table definitions at table creation to ensure data integrity from the start.

OlexTul committed:
- added Pageable to getScenariosResult() in ScenarioSourceController.java and springdoc

OlexTul committed:
- added StepRequest.java and removed Scenario.java

OlexTul committed:
- divided into 2 classes BeanConfiguration.java

Sokol3783 reviewed (changes requested):
There is a potential deadlock in this method, redo it

in parallelflowexcutorservice

OlexTul committed:
- added ScenarioService.java;
- added ScenarioServiceTest.java;
- refactored ScenarioExecutorImplTest.java;
- added ScenarioResultExceptionsTest.java, ProxySourceQueueHandlerTest.java

OlexTul committed:
- added exit the program

OlexTul commented:
> There is a potential deadlock in this method, redo it
> 
> in parallelflowexcutorservice

I made this decision - https://github.com/GFLCourses5/worker/commit/0ce8a28bd41506678110e68acae41276785973df.

OlexTul committed:
- added List<ScenarioResultResponse> getScenariosResult(@PathVariable Integer userId) to ScenarioSourceController.java

OlexTul committed:
- added proxies Cache to ProxySourceQueueHandler.java

Sokol3783 commented:
Create new branch for future commits, except cases when it's hotfix. It's uncomfortable to look for new features

OlexTul commented:
At one of the previous meetings, Artem recommended adding corrections by commits to Pull-Request. I suggest you follow his recommendation. For convenience, you can use the correspondence here, the tab https://github.com/GFLCourses5/worker/pull/125/commits, as well as the name of the commit.

For example:
![image](https://github.com/GFLCourses5/worker/assets/140727960/ec52d07a-c33a-4284-8bc7-98ab832f9807)
![image](https://github.com/GFLCourses5/worker/assets/140727960/5a93d46a-28e6-4c8c-beaf-622d3a24e10e)

OlexTul committed:
- refactoring of ScenarioResultServiceImpl.java (added options for correct removal of Step from the database);
- dropped scenarios.json, proxies.json

Sokol3783 commented:
> At one of the previous meetings, Artem recommended adding corrections by commits to Pull-Request. I suggest you follow his recommendation. For convenience, you can use the correspondence here, the tab https://github.com/GFLCourses5/worker/pull/125/commits, as well as the name of the commit.
> 
> For example: ![image](https://user-images.githubusercontent.com/140727960/277093876-ec52d07a-c33a-4284-8bc7-98ab832f9807.png) ![image](https://user-images.githubusercontent.com/140727960/277093846-5a93d46a-28e6-4c8c-beaf-622d3a24e10e.png)

[- added proxies Cache to ProxySourceQueueHandler.java](https://github.com/GFLCourses5/worker/pull/125/commits/6a31fcfd1e6f9fd66b20269bea32e16a0b334576)

and what does this  commit correct?

OlexTul commented:
6a31fcfd - added proxies Cache to ProxySourceQueueHandler.java - this commit to add Cache to ProxySourceQueueHandler.java, this improvement is not related to the comments indicated in the conversation, and as a result is not tied to them, but it is so small that it does not require a separate Pull-Request.

Sokol3783 commented on src/main/java/executor/service/service/impl/scenario/ScenarioResultServiceImpl.java:
```diff
@@ -0,0 +1,194 @@
+        deleteSteps(stepResults);
+        scenarioResultRepository.deleteById(scenarioId);
+    }
+
+    private Set<StepResult> deleteStepResults(Integer scenarioId) {
+        Set<StepResult> stepResults = getStepResults(scenarioId);
+        stepResultRepository.deleteAll(stepResults);
+        return stepResults;
+    }
+
+    private Set<StepResult> getStepResults(Integer scenarioId) {
+        return getScenarioResult(scenarioId).getStepsResults();
+    }
+
+    private ScenarioResult getScenarioResult(Integer scenarioId) {
+        return scenarioResultRepository.findById(scenarioId)
+                .orElseThrow(() -> ScenarioResultExceptions.scenarioNotFound(scenarioId));
+    }
+
+    private void deleteSteps(Set<StepResult> stepResults) {
+        stepRepository.deleteAll(getStepsForDelete(stepResults));
+    }
+
+    private List<Step> getStepsForDelete(Set<StepResult> stepResults) {
+        List<Step> stepsForDelete = new ArrayList<>();
+
+        Set<Step> steps = stepResults.stream()
+                .map(StepResult::getStep)
+                .collect(Collectors.toSet());
+
+        List<StepResult> stepResultsByStep;
+        for (Step step : steps) {
+            stepResultsByStep = stepResultRepository.findAllByStep(step);
+            if (stepResultsByStep.size() == 0) {
+                stepsForDelete.add(step);
+            }
+        }
+        return stepsForDelete;
+    }
+}
```
Do you think to simplify  this  class?
You can organize the logging using only two tables with one-to-many relationship: scenarios and actions.
In action you can add four or five fields like: 
scenario, action, value, result

I think it will give almost all information what necessary to client.

OlexTul replied:
I don't agree with you, we need to have a DataBase in third normal form.

Sokol3783 replied:
And why my variant isn't third normal form?

Sokol3783 commented on src/main/java/executor/service/controller/ScenarioSourceController.java:
```diff
@@ -0,0 +1,93 @@
+     *
+     * @param resultId The identifier of the scenario result to be deleted.
+     */
+    @DeleteMapping(value = "/{resultId}")
```
In your case any user can delete any result, fix it

OlexTul replied:
You're right, we don't have authorization on the server. I'll add a user ID

git-alexbondarenko committed:
- minor changes in pom file;
- add feign autoconfiguration;
- changes in file permissions of log file;

git-alexbondarenko committed:
- add endpoint to scenario controller to get scenario by id;
- refactor path of getting scenario results by user id;
- change id type to Long

git-alexbondarenko committed:
- refactor testShutdown() in ParallelFlowExecutorServiceTest;
- refactor WebDriverInitializerTest;
